include/wlr/types/wlr_keyboard.h

```diff
@@ -96,6 +96,7 @@ enum wlr_key_state {
 
 struct wlr_event_keyboard_key {
 	uint32_t time_msec;
+	uint32_t time_frac_nsec;
```

Comment:
For naming I think this would be cleaner as `time_nsec` with a comment explaining. If not, prefer `time_nsec_part` over `time_frac_nsec`.

Reply:
I would have preferred `time_nsec` too if it wasn't begging to be confused for `time_msec`, both semantically and spelling-wise. I chose `frac` because the fractional part of a floating point number has a very well established meaning.

Reply:
TBH I'm not a fan of splitting the timestamp like this. Using a single `uint64_t time_nsec` would be cleaner.

Reply:
Considering most event timestamps are in millisecond precision, that would add a lot of unnecessary arithmetic IMO.

Reply:
Let's go with two fields:

```c
uint32_t time_msec;
uint64_t time_nsec;
```

Which both represent the same timestamps, and don't need to be combined to get the full picture. I think this is the cleanest approach for this API, and is backwards compatible.